core/imt/src/RTaskArena.hxx

```diff
@@ -0,0 +1,143 @@
+#include <thread>
+#include <sys/stat.h>
+#include "tbb/task_arena.h"
+
+namespace ROOT {
+namespace Internal {
+
+////////////////////////////////////////////////////////////////////////////////
+/// Wrapper for tbb::task_arena.
+///
+/// tbb::task_arena is an alias of tbb::interface7::task_arena, which doesn't allow
+/// to forward declare tbb::task_arena. To avoid code breaking on tbb interface changes
+/// we don't forward declare tbb::interface7::task_arena and instead we wrap tbb::task_arena
+/// in the forward declared class RTaskArenaWrapper
+////////////////////////////////////////////////////////////////////////////////
+class RTaskArenaWrapper {
+public:
+   tbb::task_arena &Access() { return fTBBArena; }
+   Int_t TaskArenaSize() { return fTBBArena.is_active()? fTBBArena.max_concurrency() : 0; }
+   template<typename F>
+   auto Execute(F func) -> typename std::result_of<F()>::type { return fTBBArena.execute(func); }
+private:
+   tbb::task_arena fTBBArena{};
+};
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// Returns the available number of logical cores.
+///
+///  - Checks if there is CFS bandwidth control in place (linux, via cgroups,
+///    assuming standard paths)
+///  - Otherwise, returns the number of logical cores provided by
+///    std::thread::hardware_concurrency()
+////////////////////////////////////////////////////////////////////////////////
+static Int_t LogicalCPUBandwithControl()
+{
+#ifdef R__LINUX
+   // Check for CFS bandwith control
+   std::ifstream f;
+   std::string quotaFile("/sys/fs/cgroup/cpuacct/cpu.cfs_quota_us");
+   struct stat buffer;
+   // Does the file exist?
+   if(stat(quotaFile.c_str(), &buffer) == 0) {
+      f.open(quotaFile);
+      float cfs_quota;
+      f>>cfs_quota;
+      f.close();
+      if(cfs_quota > 0) {
+         std::string periodFile("/sys/fs/cgroup/cpuacct/cpu.cfs_period_us");
+         f.open(periodFile);
+         float cfs_period;
+         f>>cfs_period;
+         f.close();
+         return static_cast<int>(std::ceil(cfs_quota/cfs_period));
+      }
+   }
+#endif
+   return std::thread::hardware_concurrency();
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Factory function returning a shared pointer to the instance of the global
+//RTaskArenaWrapper.
+//
+// Allows for reinstantiation of the global RTaskArenaWrapper once all the
+//references to the previous one are gone and the object destroyed.
+////////////////////////////////////////////////////////////////////////////////
+static std::shared_ptr<ROOT::Internal::RTaskArenaWrapper> GetGlobalTaskArena()
+{
+   static std::weak_ptr<ROOT::Internal::RTaskArenaWrapper> weak_GTAWrapper;
+
+   if (weak_GTAWrapper.expired()) {
+      std::shared_ptr<ROOT::Internal::RTaskArenaWrapper> shared_GTAWrapper(new ROOT::Internal::RTaskArenaWrapper());
+      weak_GTAWrapper = shared_GTAWrapper;
+      return weak_GTAWrapper.lock();
```

Comment:
```suggestion
      return shared_GTAWrapper;
```

Reply:
Why? It's the same and it's more coherent with what I return on the other case. Makes it easier to understand.